### PR TITLE
Add more types to synapse.storage.database.

### PR DESCRIPTION
**Pull request by clokep**
I realized a few of the return types in `synapse.storage.database` were incorrect. I figured while I was checking these I should just add type hints.

### CODE REVIEW

File: synapse/storage/databases/main/ui_auth.py [hunk 1]
```diff
@@ -214,14 +215,16 @@ async def set_ui_auth_session_data(self, session_id: str, key: str, value: Any):
...
+        self, txn: LoggingTransaction, session_id: str, key: str, value: Any
+    ):
         # Get the current value.
         result = self.db_pool.simple_select_one_txn(
             txn,
             table="ui_auth_sessions",
             keyvalues={"session_id": session_id},
             retcols=("serverdict",),
-        )
+        )  # type: Dict[str, Any]  # type: ignore
```
Comment (clokep) [inline synapse/storage/databases/main/ui_auth.py:227]: See the comment on `simple_select_one_txn` for why this is necessary.

File: synapse/storage/database.py [hunk 2]
```diff
@@ -1221,40 +1302,51 @@ def simple_update_txn(txn, table, keyvalues, updatevalues):
...
-            updatevalues : dict giving column names and values to update
-            retcols : optional list of column names to return
-
-        If present, retcols gives a list of column names on which to perform
-        a SELECT statement *before* performing the UPDATE statement. The values
-        of these will be returned in a dict.
-
-        These are performed within the same transaction, allowing an atomic
-        get-and-set.  This can be used to implement compare-and-set by putting
-        the update column in the 'keyvalues' dict as well.
```
Comment (clokep) [inline synapse/storage/database.py:1241]: I'm not sure when this behavior disappeared, but `retcols` isn't a parameter anymore?

File: synapse/storage/database.py [hunk 3]
```diff
@@ -1457,21 +1574,22 @@ def simple_select_list_paginate_txn(
...
-            keyvalues (dict[str, T] | None):
+            keyvalues:
                 column names and values to select the rows with, or None to not
                 apply a WHERE clause.
-            order_direction (str): Whether the results should be ordered "ASC" or "DESC".
+            order_direction: Whether the results should be ordered "ASC" or "DESC".
+
         Returns:
-            defer.Deferred: resolves to list[dict[str, Any]]
+            The result as a list of dictionaries.
```
Comment (clokep) [inline synapse/storage/database.py:1592]: This was incorrect, this is a synchronous function that just returns a list of dictionaries.

File: synapse/storage/database.py [hunk 4]
```diff
@@ -1516,19 +1641,26 @@ def simple_search_list(self, table, term, col, retcols, desc="simple_search_list
...
-            retcols (iterable[str]): the names of the columns to return
+            txn: Transaction object
+            table: the table name
+            term: term for searching the table matched to a column.
+            col: column to query term should be matched to
+            retcols: the names of the columns to return
+
         Returns:
-            defer.Deferred: resolves to list[dict[str, Any]] or None
+            0 if no term is given, otherwise a list of dictionaries.
```
Comment (clokep) [inline synapse/storage/database.py:1663]: This is not a Deferred, it is synchronous.

**Comment by clokep** [comment]
This is vaguely related to #8119 in that it makes some of the types clearer, but isn't really part of that work.

File: synapse/storage/database.py [hunk 5]
```diff
@@ -1516,19 +1643,26 @@ def simple_search_list(self, table, term, col, retcols, desc="simple_search_list
...
     @classmethod
-    def simple_search_list_txn(cls, txn, table, term, col, retcols):
+    def simple_search_list_txn(
+        cls,
+        txn: LoggingTransaction,
+        table: str,
+        term: Optional[str],
+        col: str,
+        retcols: Iterable[str],
+    ) -> Union[List[Dict[str, Any]], int]:
```
Comment (clokep) [inline synapse/storage/database.py:1653]: I wonder if this should return `Optional[List[Dict[str, Any]]]` instead? This was what the old comment said, but definitely isn't what the code does. 😕

Comment (erikjohnston) [reply synapse/storage/database.py:1653]: Ah, I assume that it probably works as `0` is also falsey. Let's not try and fix it in this PR though.

**Review by erikjohnston — APPROVED** [review]

File: synapse/storage/database.py (hunk 5 above)
Comment (clokep) [reply synapse/storage/database.py:1653]: That was my assumption too. I can do a follow-up that investigates a bit more closely.

File: synapse/storage/database.py [hunk 6]
```diff
@@ -387,8 +390,15 @@ def loop():
...
-    ):
+        self,
+        conn: Connection,
+        desc: str,
+        after_callbacks: List[_CallbackListEntry],
+        exception_callbacks: List[_CallbackListEntry],
+        func: Callable,
+        *args: Any,
+        **kwargs: Any
+    ) -> Any:
```
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/storage/database.py]: Done! 👍

**Pull request merged**